src/broadcast_service.rs

```diff
@@ -32,160 +31,110 @@ pub enum BroadcastServiceReturnType {
+            .collect();
 
-    let blobs_chunking = Instant::now();
-    // We could receive more blobs than window slots so
-    // break them up into window-sized chunks to process
-    let window_size = window.read().unwrap().window_size();
-    let blobs_chunked = blobs_slot_heights
-        .chunks(window_size as usize)
-        .map(|x| x.to_vec());
-    let chunking_elapsed = duration_as_ms(&blobs_chunking.elapsed());
+        // TODO: blob_index should be slot-relative...
+        index_blobs(&blobs, &self.id, self.blob_index, &slots);
 
-    let broadcast_start = Instant::now();
-    for blobs in blobs_chunked {
-        let blobs_len = blobs.len();
-        trace!("{}: broadcast blobs.len: {}", id, blobs_len);
+        let to_blobs_elapsed = duration_as_ms(&to_blobs_start.elapsed());
 
-        index_blobs(blobs.iter(), &node_info.id, *receive_index);
+        let broadcast_start = Instant::now();
 
-        // keep the cache of blobs that are broadcast
         inc_new_counter_info!("streamer-broadcast-sent", blobs.len());
-        {
-            let mut win = window.write().unwrap();
-            assert!(blobs.len() <= win.len());
-            let blobs: Vec<_> = blobs.into_iter().map(|(b, _)| b).collect();
-            for b in &blobs {
-                let ix = b.read().unwrap().index().expect("blob index");
-                let pos = (ix % window_size) as usize;
-                if let Some(x) = win[pos].data.take() {
-                    trace!(
-                        "{} popped {} at {}",
-                        id,
-                        x.read().unwrap().index().unwrap(),
-                        pos
-                    );
-                }
-                if let Some(x) = win[pos].coding.take() {
-                    trace!(
-                        "{} popped {} at {}",
-                        id,
-                        x.read().unwrap().index().unwrap(),
-                        pos
-                    );
-                }
 
-                trace!("{} null {}", id, pos);
-            }
-            for b in &blobs {
-                {
-                    let ix = b.read().unwrap().index().expect("blob index");
-                    let pos = (ix % window_size) as usize;
-                    trace!("{} caching {} at {}", id, ix, pos);
-                    assert!(win[pos].data.is_none());
-                    win[pos].data = Some(b.clone());
-                }
-            }
+        db_ledger
+            .write_consecutive_blobs(&blobs)
+            .expect("Unrecoverable failure to write to database");
 
-            db_ledger
-                .write_consecutive_blobs(&blobs)
-                .expect("Unrecoverable failure to write to database");
-        }
+        // don't count coding blobs in the blob indexes
+        self.blob_index += blobs.len() as u64;
+
+        // Send out data
+        ClusterInfo::broadcast(&self.id, last_tick, &broadcast_table, sock, &blobs)?;
```

Comment:
Do we care about retrying a broadcast if it fails?  The old window code wouldn't increment transmit_index.data https://github.com/solana-labs/solana/pull/2389/files#diff-35fd7e6bd55c89849b6fac670db810cdL540 on detecting the first transmission failure, so that broadcast would retry a failure on the next iteration.

Reply:
I think we care, but don't have an easy facility to implement right now.